cmd/schema_gen.go

```diff
@@ -0,0 +1,195 @@
+			if err != nil {
+				fmt.Println("Error writing YML to file in path")
+				os.Exit(1)
+			}
+			fmt.Println("\n✅ " + modelName + "schema successfully updated at path: " + ymlPath)
+		}
+	},
+}
+
+func marshalYML(schema interface{}) string {
+	y, err := yaml.Marshal(&schema)
+	if err != nil {
+		fmt.Println("Could not marshal schema")
+	}
+	yml := fmt.Sprintf("\n%s\n\n", string(y))
+	return yml
+}
+
+func GetColumnsForModel(modelName string, target *config.Target) ([]string, error) {
+	schema, err := bigquery.GetColumnsFromTable(modelName, target)
+	if err != nil {
+		return nil, err
+	}
+
+	columns := []string{}
+	for _, FieldSchema := range schema {
+		column := fmt.Sprintf("%v", FieldSchema.Name)
+		columns = append(columns, column)
+	}
+	return columns, nil
+}
+
+func WriteYMLToFile(filePath string, body string) error {
+	f, err := os.Create(filePath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	_, _ = f.WriteString("# Auto-generated by DDBT at " + time.Now().String())
+	_, err = f.WriteString(body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// generate an empty schema file which will be populated according to existing yml schemas and the bigquery schema.
+// Returns the local path for the yml file and the yml file struct
+func generateEmptySchemaFile(model *fs.File) (ymlPath string, schemaFile properties.File) {
+	ymlPath = strings.Replace(model.Path, ".sql", ".yml", 1)
+	schemaFile = properties.File{}
+	schemaFile.Version = properties.FileVersion
+	return ymlPath, schemaFile
+}
+
+// generate a new schema model for the provided model name and bqcolumns
+// this is used when then is no existing model
+func generateNewSchemaModel(modelName string, bqColumns []string) properties.Model {
+	schemaModel := properties.Model{}
+	schemaModel.Name = modelName
+	schemaModel.Description = "Please fill this in with a useful description.."
+	schemaModel.Columns = []properties.Column{}
+	for _, bqCol := range bqColumns {
+		column := properties.Column{}
+		column.Name = bqCol
+		schemaModel.Columns = append(schemaModel.Columns, column)
+	}
+
+	return schemaModel
+}
+
+// check if bq column is in schema (add missing)
+func addMissingColumnsToSchema(schemaModel *properties.Model, bqColumns []string) {
+	columnsAdded := []string{}
+	for _, bqCol := range bqColumns {
+		columnFound := FindColumnInSchema(bqCol, schemaModel)
+		if !columnFound {
+			column := properties.Column{}
+			column.Name = bqCol
+			schemaModel.Columns = append(schemaModel.Columns, column)
+			columnsAdded = append(columnsAdded, bqCol)
+		}
+	}
+	fmt.Println("➕ Columns added to Schema (from BQ table):", columnsAdded)
+}
+
+// check if schema column in bq (remove missing)
+func removeOutdatedColumnsFromSchema(schemaModel *properties.Model, bqColumns []string) {
+	columnsRemoved := []string{}
+	columnsKept := properties.Columns{}
+	for _, schemaCol := range schemaModel.Columns {
+		columnFound := FindSchemaColumnInSlice(schemaCol, bqColumns)
```

Comment:
Small tweak that will help you avoid the nested `for` loop:
- First turn `*schemaModel` into a `map[string]*schemaModel` (before this loop), to map column names to schema models
- Then you can just to a lookup inside of the map, instead of using `FindColumnInSchema()` which iterates over the whole schema

Reply:
Nice! Definitely feel like I learned to do that in a course and forgot it 😂  Have added! Looks much cleaner - thanks!